README
$Header$

What it does: See file DESCRIPTION
How to use:   See file example/proxy.sh

To compile, just type
	make

This now is based on tinolib to improve further development.

Additional notes:
	The Apache developers wrote a good source about lingering:
	http://httpd.apache.org/docs/misc/fin_wait_2.html
	This is true for HTTPD but is true for many other cases as well.

-Tino, webmaster@scylla-charybdis.com
$Log$
Revision 1.2  2005-08-19 04:26:39  tino
release socklinger 1.3.0

Revision 1.1  2004/08/16 01:29:57  Administrator
initial add


ChangeLog
2010-01-26  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger): max linger time implemented
	(process_args): Option -w

2010-01-25  Valentin Hilbig  <tino@firebird.03.softkill.net>

	* socklinger.c (process_args): usage corrected
	(process_args): Option -f
	(socklinger_dosock): connect() uses tino_proc_sock() to support
	'|script' sockets

2009-07-17  Valentin Hilbig  <tino@zoo.03.softkill.net>

	* socklinger.c (process_args, socklinger): PID moved from
	informational to environment

2009-02-04  Valentin Hilbig  <tino@zoo.03.softkill.net>

	* socklinger.c (conf_set_time): added
	(process_args): Options -p and -u added
	(drop_privileges): added, notes privilege dropping errors
	(main): in case of -c without -b drop privileges early

2007-05-08    <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger): better lingering and shutdown now
	from tinolib
	(vnote): added to better generalize routines
	(socklinger_error): moved to top

2007-04-29    <tino@firebird.03.softkill.net>

	* socklinger.c (main): SIGCHLD now is delivered to the process and
	more commandline parameters

2007-04-04    <tino@firebird.03.softkill.net>

	* socklinger.c (process_args): future options prepared

	* example/tester.sh: improved output for tests

	* socklinger.c (socklinger_postfork): cosmetics: again typo in PID
	printing removed (now in waitchild) and SOCKLINGER_NR was 0..n-1,
	correct to 1..n (now based on child_findfree).
	(socklinger_dosock): sleep replaced by tino_relax()
	(socklinger_error): added for support of option -1
	(socklinger_child_findfree, socklinger_waitchild)
	(socklinger_forkchild): added as generalized parts for pre- and
	postfork
	(socklinger_prefork, socklinger_postfork): changed to use new
	routines (which support ignerr)
	(socklinger_forkchild): set dodelay only if delay is active

2007-04-03    <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger_postfork, socklinger_prefork): delay
	added (now in waitchild())

2007-04-02    <tino@firebird.03.softkill.net>

	* socklinger.c (process_args): delay added

2006-12-12    <tino@firebird.03.softkill.net>

	* BUGS: updated

	* diet: Softlink deleted, no "make static"

	* Makefile.tino: re-created with tino/setuptino.sh

	* example/portcullis.sh: completed (hopefully)

2006-09-16  U-KOSH\Administrator  <Administrator@KOSH>

	* socklinger.c (main): bugfix: if @ was missing (like in IPs), cnt
	was set.

2006-09-08  U-KOSH\Administrator  <Administrator@KOSH>

	* socklinger.c (socklinger_prefork, socklinger_postfork): added as
	CygWin fix.  Change usage such, that '-n@' gives alternate fork
	method and '>' creates the connect variant.
	
2006-06-17    <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger): SOCKLINGER_SOCK added to environment
	(socklinger): Now correct exist status is printed

2006-02-09    <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger): close socket on forked connection
	(main): additional socket close here, too.

2006-01-25    <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger_accept): HUP control added

2006-01-24    <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger_accept): If SYN/RST was found or
	something similar weird happened, then accept returns "broken
	pipe" and socklinger got an error.  Now the process relaxes a
	little bit and retries.
	(socklinger): whoopsie, errno was checked even if there was no
	error ..

2005-10-30    <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger): improved output to screen

	* example/streamgrab.sh: added

	* example/tester.sh: improved to give more info

2005-03-15    <tino@firebird.03.softkill.net>

	* socklinger.c (socklinger): fo now specially treated if fo!=fi

2004-08-16  U-KOSH\Administrator  <Administrator@KOSH>

	* socklinger.c (socklinger_accept, main): prefix "count@" for
	number of parallel connections implemented.  It looks like it is
	preforking, but the script isn't preforking at all.  Socklinger
	terminates as soon as any child comes back, but the running childs
	are not terminated yet by force.


example/portcullis.sh
#!/bin/sh
# $Header$
#
# Portcullis Script for socklinger.
# Use this to protect a service.
#
# Example:
#	( cd ..; make install; )
#	/usr/local/bin/socklinger 10@:23 /usr/local/sbin/portcullis.sh '[ .127.0.0.1 = ".$IP" ] || echo Password' /usr/sbin/sshd -i
#
# $Log$
# Revision 1.1  2006-12-12 10:36:42  tino
# See ChangeLog
#

TMP=/tmp
ALLOW=3600
BLOCK=60
if [ 0 = $# ]
then
	echo "Usage: `basename "$0"` 'checkscript args' command args.." >&2
	echo "	If permission is still valid (timestamp), peer is permitted." >&2
	echo "	Else checkscript is invoked." >&2
	echo "	If checkscript returns FALSE access is denied." >&2
	echo "	If checkscript returns a text, a passord prompt is done." >&2
	echo "	The password must be entered successfully to become permitted." >&2
	echo "	If checkscript returns empty string, peer is permitted." >&2
	echo "	If permitted, the command is invoked with args." >&2
	echo "	By default a permission is granted for $ALLOW seconds." >&2
	echo "	A wrong password penalty is $BLOCK seconds." >&2
	echo "	Checkscript can utilize $IP and $STAMP (filename)." >&2
	exit 1
fi

log()
{
printf '[%s][%05d] - %s\n' "$SOCKLINGER_NR" "$PPID" "$*" >&2
}

ex()
{
log "$*"
exit 1
}

no()
{
echo NO
ex "NOT permitted peer $IP: $*"
}

[ -z "$SOCKLINGER_PEER" ] &&
ex "Only use this from socklinger!"

IP="$SOCKLINGER_PEER"
case "$SOCKLINGER_PEER" in
*/* | *..*)	ex "illegal peer $SOCKLINGER_PEER";;
\[*\]:*)	IP="`expr "$IP" : '\[\(.*\)\]:[^:]*$'`";;
*:*)		IP="`expr "$IP" : '\(.*\):[^:]*$'`";;
esac

# Calculate timestamp window in which everything is OK
NOW="`date +%s`"
MAX="`expr "$NOW" + "$ALLOW"`"

# Hunt for the peer name
DIR="$TMP/portcullis/$LOGNAME/sock-$SOCKLINGER_SOCK"
STAMP="$DIR/peer-$IP.stamp"

# Remove permission if stale
# And set the deny flag according to permission
deny=""
[ -s "$STAMP" ] &&
if	! read valid deny < "$STAMP" ||
	[ "$NOW" -gt "$valid" ] ||
	[ "$MAX" -lt "$valid" ]
then
	if rm -f "$STAMP" 2>/dev/null
	then
		log "stale permission for peer $IP"
		rmdir "$DIR" 2>/dev/null
	else
		log "cannot revoke stale permission for peer $IP"
		deny="stale permission"
	fi
fi

if [ -s "$STAMP" ]
then
	[ -n "$deny" ] && no "$deny"
	log "permission valid for peer $IP"
else
	# Run the checkscript:
	#
	# This script returns:
	# 1			PEER NOT ALLOWED
	# 0 + empty string:	peer allowed
	# 0 + some STRING:	ask for this PASSWORD
	export IP STAMP
	try="`/bin/sh -c "$1"`" || no "permission denied"

	if [ -n "$try" ]
	then
		echo "Enter password:"
		read -t10 pw || no "timeout"
		if [ ".$try" != ".$pw" ] && [ ".$try" != ".$pw" ]
		then
			[ 0 -lt "$BLOCK" ] && echo "`expr "$NOW" + "$BLOCK"` blocked" > "$STAMP"
			no "invalid password"
		fi
		mkdir -p "$DIR" || no "cannot create dir"
		chmod 700 "$DIR" || no "cannot set dir mode"
		echo "$MAX" > "$STAMP" &&
		echo OK
 		log "permission granted to peer $IP"
		# Now throw out the client,
		# next time it gets the service.
		exit 0
	fi
	# Fall through service
	log "directly permitted peer $IP"
fi

shift
exec "$@"


example/proxy.sh
#!/bin/bash
#
# $Header$
#
# WARNING!  THIS SCRIPT WAS DONE IN A HURRY!
#
# This is an extremely simple proxy for network based DEBIAN install.
# It can be started on demand and caches all files downloaded, so they
# need not be downloaded again.
# It can run on another Linux machine or under Windows running Cygwin.
# It is able to restart downloads from where they broke.
#
# Usage:
#	make
#	mkdir -p PROXY
#	rm -rf PROXY/head PROXY/tmp
#	./socklinger -sn-2 :8081 example/proxy.sh [proxy proxyport [Proxyline]]&
#	SP="$!"
#	trap 'kill $SP' 0
#	tail --retry --follow=name PROXY/LOG PROXY/MIST
#
# Then tell Debian to use a proxy on port 8081 of the machine
# where socklinger runs to download Debian packages.
# Note that you must allow 2 parallel connects, as else
# the Debian Installer starves at one location (bug in d-i?).
#
# Note on PROXY, PROXYPORT and PROXYLINE:
#
# You can hand a proxy to this script.  The Proxyline is a line which is
# transparently sent to to Proxy.  You can use this for user authentication.
# This can be done with a line like:
# "Proxy-Authorization: Basic B64STRING"
# where B64STRING is the base64 encoded string of "user:password".
# To calculate this string you can try something like:
# php -r 'echo base64_encode("user:password");'; echo
#
# You need a directory named PROXY where the data is written to.
# In PROXY/LOG a log is written.
# In PROXY/MIST the errors are written ("Mist" in German means Bullshit).
#
# THIS SCRIPT HAS A LOT OF BUGS:
#
# It has a lot of race conditions.  So never allow more than one Debian
# to download new packages via this script in parallel!
#
# It is only working with Debian install and apt-get.
# You can only access mirrors in the domain debian.org.
# See ### ADD ALLOWED DOMAINS HERE ### to enter more alloed domains.
#
# It does not check for changes on the Debian mirror!  To get new files,
# remove the header files (rm -rf PROXY/head).
# YOU HAVE TO DO THIS MANUALLY OR BY CRON!!!
#
# $Log$
# Revision 1.3  2008-01-24 23:42:38  tino
# Improved for proxy support etc.
#
# Revision 1.2  2004-09-20 01:52:14  tino
# Clarified, you must allow 2 proxies running in parallel.

cd PROXY || exit 1
umask 002

mkdir -p tmp

PROXY="$1"
PROXYPORT="$2"
PROXYLINE="$3"

#exec 2>>LOG

mist()
{
{
echo "==============================="
echo "`date` MIST: $*"
echo "host=$host"
echo "port=$port"
echo "$get $url $rest"
while read -t10 line
do
	echo "$line"
done
} >> MIST
exit 1
}

read -r -t 20 get url rest || mist "EOF" 
[ ".GET" != ".$get" ] && mist "no get"
case "$url" in
*)	url="`expr "$url" : '\(.*\).$'`";;
esac

case "$url" in
*..*)	mist ".. im URL";;
http://*)	;;
*)	mist "seltsames URL";;
esac

host="`expr "$url" : 'http://\([0-9a-z.]*\)[:/]'`"
port="`expr "$url" : 'http://[^:/]*:\([0-9]*\)/'`"

[ -z "$port" ] && port=80
[ -z "$host" ] && mist "host?"
case "$host" in
*.debian.org)	;;
### ADD ALLOWED DOMAINS HERE ###
*)		mist "no debian";;
esac

fhost="$host"
fport="$port"
[ -z "$PROXY" ] || fhost="$PROXY"
[ -z "$PROXYPORT" ] || fport="$PROXYPORT"

: fetch
fetch()
{
{
echo "$1 $url HTTP/1.0"
if [ 80 != "$port" ]
then
	echo "Host: $host:$port"
else
	echo "Host: $host"
fi
[ -z "$PROXYLINE" ] || echo "$PROXYLINE"
echo ""
} |
netcat -w60 "$fhost" "$fport"
}

spcparm()
{
eval $1='`expr "$line" : "[^:]*:[[:space:]]*\\([^[:space:]]*\\)"`'
}

fullparm()
{
eval $1='`expr "$line" : "[^:]*:[[:space:]]*\\(.*\\)\\$"`'
}

: parsehead
parsehead()
{
read response code rest || return 1
[ 200 = "$code" ] || return 1
while	line=""
	read -r line || return 1
	[ -n "$line" -a "" != "$line" ]
do
	case "$line" in
	[Cc]ontent-[Tt]ype:*)	spcparm $1type;;
	[Cc]ontent-[Ll]ength:*)	spcparm $1len;;
	[Ll]ast-[Mm]odified:*)	fullparm $1date;;
	esac
done
return 0
}

: cathead
cathead()
{
while	read -r line || exit 1
	[ -n "$line" -a "" != "$line" ]
do
	echo "$line"
done
echo "$line"
}

: chklen
chklen()
{
clen="`{ parsehead c; [ ":$ctype:$clen:$cdate" = ":$type:$len:$date" ] && wc -c; }`" || return 1
[ -z "$len" -o ".$len" = ".$clen" ]
}

durl="$url"
furl="$url"
case "$url" in
*/)	furl="$url""index.html";;
*)	durl="`dirname "$url"`";;
esac

if [ -f "data/$furl" ]
then
	if [ ! -s "head/$furl" ]
	then
		mkdir -p "head/$durl" || exit 1
		echo "++ $url" >> LOG
		fetch HEAD > "head/$furl"
	fi

	len=""
	if	{ ! [ "head/$furl" -nt "data/$furl" ]; } ||
		{ parsehead < "head/$furl" &&
		  chklen < "data/$furl" &&
		  touch -r "data/$furl" "head/$furl";
		}
	then
		echo "-- $url" >> LOG
		exec cat "data/$furl"
	fi

	echo "up $url" >> LOG
	mv -f --backup=t "head/$furl" "head/$furl.old"
fi

echo "do $url" >> LOG

TMP="tmp/$$"
fetch GET | tee "$TMP"

[ -f "$TMP" ] || exit 1
parsehead < "$TMP" || exit 1
chklen < "$TMP" || exit 1

echo "ok $url" >> LOG

mkdir -p "head/$durl" || exit 1
cathead < "$TMP" > "head/$furl"

mkdir -p "data/$durl" || exit 1
mv -f --backup=t "$TMP" "data/$furl"

